### PR TITLE
[Config] In XmlUtils, avoid converting from octal every string starting with a 0

### LINKED ISSUE #41532: Convert XML to PHP array  with XMLUtils::convertDomElementToArray : numbers who start by "0" convert into octal number (which is wrong)
**Symfony version(s) affected**: 5.2.5 (at least)

**Description**
If you have some value in your XML file with a number starting by "0" (example : <ArrivalTime>080000</ArrivalTime> ), XMLUtils::convertDomElementToArray will return "ArrivalTime" => 0, which isn't correct.

**How to reproduce**

Let's take the following XML file :

`$xml_file = <?xml version="1.0"?><ResponseItinerary><Header><CruiseLineCode>SBN</CruiseLineCode><SubsystemId>3</SubsystemId><AgencyId1>58080</AgencyId1><AgencyId2>58080</AgencyId2><Currency/><AgencyConsumer/><TransactionCounter/></Header><Itinerary><GeneralInfo><Errors/><Warnings/><HowMuch>29</HowMuch></GeneralInfo><List ItineraryCode="E1N12BVN" PackageId="V175"><ListElement><DWeek>SAMEDI</DWeek><DepartureDate>11/12/2021</DepartureDate><PortName Code="GRW" Activity="">GREENWICH (LONDON)</PortName><ArrivalTime>090000</ArrivalTime><DepartureTime>000000</DepartureTime><Indicator/></ListElement></List></Itinerary></ResponseItinerary>;`

Let's convert it into PHP file using XmlUtils Symfony class : 

```
$oDom = new \DOMDocument;
$oDom->preserveWhiteSpace = false;
$oDom->loadXML($xml_file);
$oDom->formatOutput = true;
  
$xml_list_element = $oDom->getElementsByTagName('List')->item(0); 

$content = self::convertDomElementToArray($content);
dump($content);

// Here is the dump 
/*
array:3 [
  "ItineraryCode" => "E1N12BVN"
  "PackageId" => "V175"
  "ListElement" => array:16 [
    0 => array:6 [
      "DWeek" => "SAMEDI"
      "DepartureDate" => "11/12/2021"
      "PortName" => array:3 [
        "Code" => "GRW"
        "Activity" => ""
        "value" => "GREENWICH (LONDON)"
      ]
      "ArrivalTime" => 0 // IT SHOULD BE "090000" 
      "DepartureTime" => 0
      "Indicator" => null
    ]
]
*/
```
The "ArrivalTime" property shouldn't show 0, but 090000.

**Possible Solution**
Here's the solution : 

1) add the following function into app/vendor/symfony/config/Util/XmlUtils.php
`private function is_octal($x) {
        return decoct(octdec($x)) == $x;
    }`
2) change a line in public static function phpize() (still into app/vendor/symfony/config/Util/XmlUtils.php)
change this
`return '0' == $value[0] ? octdec($value) : (((string) $raw === (string) $cast) ? $cast : $raw);`
into this
`return self::is_octal($value) ? octdec($value) : (((string) $raw === (string) $cast) ? $cast : $raw);`

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #41532 
| License       | MIT
| Doc PR        | N/A

One word on the use of `intval` instead of `octdec`: `octdec` leads to deprecation notices when testing a non-octal string:
```
Invalid characters passed for attempted conversion, these have been ignored
```
While `intval` doesn't, which worth the use in my opinion.

### CODE REVIEW

File: src/Symfony/Component/Config/Util/XmlUtils.php [hunk 1]
```diff
@@ -281,4 +277,9 @@ protected static function getXmlErrors($internalErrors)
 
         return $errors;
     }
+
+    protected static function isOctal(string $str): bool
```
Comment (stof) [inline src/Symfony/Component/Config/Util/XmlUtils.php]: please make it private. There is no reason to make this an extension point

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Config/Util/XmlUtils.php (hunk 1 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Config/Util/XmlUtils.php]: You're right, done! :heavy_check_mark:

File: src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php [hunk 2]
```diff
@@ -169,6 +169,8 @@ public function getDataForPhpize(): array
             [1, '1'],
             [-1, '-1'],
             [0777, '0777'],
+            ['-0777', '-0777'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: isn't this supposed to be parsed to `-0777` (aka no quotes)?

Comment (alexandre-daubois) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: Well, `decoct` has hard times with negative numbers:
```php
echo decoct(-1); // "1777777777777777777777"
```
This makes me wonder if we should support negative octal numbers. I did some research on octal notation and negative numbers, and it reinforced this idea.
If you think it worth to add a few lines to support it (like checking the strings starts with `-` and then multiply by -1 the final result, something like that), I can add it. But it seems a bit "hacky" :smile:

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: On the other end:
```php
echo -0777; // -511
```

Comment (alexandre-daubois) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: You're right! I'll come up with a better solution to handle this. 👍

Comment (alexandre-daubois) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: This should be better now!

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Pull request merged**